docs/types/types.md

```diff
@@ -1,395 +1,421 @@
-define subtyping relationships in a two-step process.
-
-First, we gather all of the type variables used in its definition and build a
-new type with those type variables expressed in sequence. Then if certain
-conditions are met (e.g. the two have the exact same shape), we allow punning
-between the basic type constructor and this elaborated type constructor.
-
-e.g.
+This creates a type `Foo` with three type arguments. For example:
 
+```ruby
+main = 42:(Foo Float Integer Text)
 ```
-type Book author title year
-```
-
-is equivalent to defining
 
-```
-type Book.type author_type title_type year_type
-  type Book (author : author_type) (title : title_type) (year : year_type)
-```
+fails with `Type error: expected expression to be Foo, but got Integer.` as `42`
+isn't value of type `Foo`.
 
-whereupon we notice that Book.type and Book are the same shape, and allow
-punning between `Book.type` and `Book`.
+One can add any number of _atom constructors_ to such a type. For example
+following code snippet
 
-```
-Book "Jane Austen" "Pride and Prejudice" 1832 : Book.type Text Text Int
-Book "Jane Austen" "Pride and Prejudice" 1832 : Book Text Text Int
+```ruby
+type Foo x y z
+    Zero
+    One u:x
+    Two u:x v:y
+    Three u:x v:y w:z
 ```
 
-The eagle-eyed among you might note that author_type and author are different
-field names, so this would actually elaborate out to
+adds four _atom constructors_ named `Foo.Zero`, `Foo.One`, etc. with fields
+named `x`, `u`, `w` initialized to provided arguments when these constructors
+are used. The type `Foo` automatically offers getters for fields `u`, `v` and
+`w`. However some of these fields may yield _undefined error value_ when other
+than `Foo.There` constructor (which assignes all the field values) is used. With
+the updated `Foo` type definition following two versions of `main` function type
+check fine:
 
+```ruby
+main = Foo.Zero:(Foo Float Integer Text)       # yields Zero
+main = (Foo.One 3.14):(Foo Float Integer Text) # yields (One 3.14)
 ```
-type Book.type author title year
-  type Book (author : author) (title : title) (year : year)
-```
-
-where we are careful with lexical scopes to distinguish between the term level
-author and the type level author, enabling this pun.
 
-In the List example
+Arguments to atom constructors can have defaults and explicit type signatures.
 
-```
+> [!WARNING]
+>
+> Following seems to be off a bit:
+>
+> Due to two things that feel like 'primes' popping up in the conversation
+> below, we've chosen to denote one with 'x.type' and the other with 'Code x'.
+>
+> However, there are some limitations on type signatures that will be discussed
+> later. There is a distinction between type and its atoms. Types themselves are
+> effectively represented as atoms and type constructors are atom constructors
+> as well.
+>
+> `List` plays multiple roles here. It acts as the type constructor which can be
```

Comment:
I guess the word you are looking for is _higher-kinded_.

Reply:
> ... you are looking for ...

Those are @ekmett [words](https://github.com/enso-org/enso/blame/c2f0925064918eb7851c5072e9f0fa0adfb63975/docs/types/types.md#L36)'s not mine! Changed in 0213941